relays/substrate/src/messages_source.rs

```diff
@@ -0,0 +1,221 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Substrate client as Substrate messages source. The chain we connect to should have
+//! runtime that implements `<BridgedChainName>HeaderApi` to allow bridging with
+//! <BridgedName> chain.
+
+use async_trait::async_trait;
+use bp_message_lane::{LaneId, MessageNonce};
+use bp_runtime::InstanceId;
+use codec::{Decode, Encode};
+use frame_support::weights::Weight;
+use messages_relay::{
+	message_lane::{MessageLane, SourceHeaderIdOf, TargetHeaderIdOf},
+	message_lane_loop::{ClientState, SourceClient, SourceClientState},
+};
+use relay_substrate_client::{Chain, Client, Error as SubstrateError, HashOf, HeaderIdOf};
+use relay_utils::HeaderId;
+use sp_core::Bytes;
+use sp_runtime::{traits::Header as HeaderT, DeserializeOwned};
+use sp_trie::StorageProof;
+use std::{marker::PhantomData, ops::RangeInclusive};
+
+/// Substrate client as Substrate messages source.
+pub struct SubstrateMessagesSource<C: Chain, P, M> {
+	client: Client<C>,
+	tx_maker: M,
+	lane: LaneId,
+	instance: InstanceId,
+	_marker: PhantomData<P>,
+}
+
+/// Substrate transactions maker.
+#[async_trait]
+pub trait SubstrateTransactionMaker<C: Chain, P: MessageLane>: Clone + Send + Sync {
+	/// Signed transaction type.
+	type SignedTransaction: Send + Sync + Encode;
+
+	/// Make messages receiving proof transaction.
+	async fn make_messages_receiving_proof_transaction(
+		&self,
+		generated_at_block: TargetHeaderIdOf<P>,
+		proof: P::MessagesReceivingProof,
+	) -> Result<Self::SignedTransaction, SubstrateError>;
+}
+
+impl<C: Chain, P, M> SubstrateMessagesSource<C, P, M> {
+	/// Create new Substrate headers source.
+	pub fn new(client: Client<C>, tx_maker: M, lane: LaneId, instance: InstanceId) -> Self {
+		SubstrateMessagesSource {
+			client,
+			tx_maker,
+			lane,
+			instance,
+			_marker: Default::default(),
+		}
+	}
+}
+
+impl<C: Chain, P, M: Clone> Clone for SubstrateMessagesSource<C, P, M> {
+	fn clone(&self) -> Self {
+		Self {
+			client: self.client.clone(),
+			tx_maker: self.tx_maker.clone(),
+			lane: self.lane,
+			instance: self.instance,
+			_marker: Default::default(),
+		}
+	}
+}
+
+#[async_trait]
+impl<C, P, M> SourceClient<P> for SubstrateMessagesSource<C, P, M>
+where
+	C: Chain,
+	C::Header: DeserializeOwned,
+	C::Index: DeserializeOwned,
+	<C::Header as HeaderT>::Number: Into<u64>,
+	P: MessageLane<
+		MessageNonce = MessageNonce,
+		MessagesProof = (Weight, (HashOf<C>, StorageProof, LaneId, MessageNonce, MessageNonce)),
```

Comment:
I think it would be worthwhile to typedef this

Reply:
Also, why does the proof have the weight included in it?

Reply:
Because weight needs to be submitted with proof.

Reply:
(I'll add a typedef, though :) )

Reply:
> Because weight needs to be submitted with proof.

I was gonna ask "why" again, but it looks like you answered it with the doc comments 👍